Make anchor positioning fallbacks respond to scrolling [css-anchor-position-1]

fantasai opened the pull request:
#### 90fc9aa7e180c7b0fb56c76a2cdb5b202a0c128a
<pre>
Make anchor positioning fallbacks respond to scrolling [css-anchor-position-1]
<a href="https://bugs.webkit.org/show_bug.cgi?id=295003">https://bugs.webkit.org/show_bug.cgi?id=295003</a>
<a href="https://rdar.apple.com/154349365">rdar://154349365</a>

Reviewed by Antti Koivisto.

This patch rewrites are scroll handling for anchor positioning to incorporate
the current scroll position into layout so that anchored boxes can respond to
space available at the current scroll position.

In order to maintain responsiveness to scrolling, it caches scroll position
snapshots with each layout, which allows us to rapidly calculate adjustments
o the anchored box&apos;s scroll adjustment transform on changes to the scroll
position outside of layout as well as to trigger more invasive invalidation
when necessary in response to scroll position changes, e.g. for fallbacks.

* LayoutTests/TestExpectations:

Pass more tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-chained-002-expected.html:

Fix test reference.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-js-expose-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt:

Pass more tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-011-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-scrolling-002.tentative-expected.txt:

Rebaseline.

* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:

Mark WK1 tests as still failing, to debug later per Antti&apos;s instructions.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::updateAnchorPositionedAfterScroll):

Update anchor scroll adjustment call site.

* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::anchorScrollAdjusterFor):
(WebCore::LocalFrameViewLayoutContext::registerAnchorScrollAdjuster):
(WebCore::LocalFrameViewLayoutContext::unregisterAnchorScrollAdjusterFor):
(WebCore::LocalFrameViewLayoutContext::invalidateAnchorDependenciesForScroller):
(WebCore::LocalFrameViewLayoutContext::removeScrollerFromAnchorScrollAdjusters):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:

Set up storage for anchor scroll adjuster objects.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):

Update anchor scroll adjustment call site.

* Source/WebCore/platform/graphics/LayoutPoint.h:
(WebCore::LayoutPointLimits::clamp):
(WebCore::LayoutPointLimits::fits):
(WebCore::LayoutPointLimits::distance):
* Source/WebCore/platform/graphics/LayoutSize.h:
(WebCore::LayoutSizeLimits::clamp):
(WebCore::LayoutSizeLimits::fits):
(WebCore::LayoutSizeLimits::distance):

Introduce some convenience objects / methods.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::captureAnchorGeometry):

Adjust call parameters to computeAnchorRectRelativeToContainingBlock().

* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::originalContainingRange const):
(WebCore::PositionedLayoutConstraints::insetModifiedContainingRange const):

Export more info.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::simplifiedLayout):
(WebCore::RenderBlock::layoutOutOfFlowBox):

Call scroll snapshot capture method after layout.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::willBeDestroyed):
(WebCore::RenderBox::styleDidChange):

Manage anchor scroll adjuster memory allocation and invalidation.

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::layoutIfNeeded):

Call scroll snapshot capture method after layout.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransformFromStyle const):
(WebCore::RenderLayer::setAnchorScrollAdjustment):
(WebCore::RenderLayer::clearAnchorScrollAdjustment):
(WebCore::RenderLayer::setSnapshottedScrollOffsetForAnchorPositioning): Deleted.
(WebCore::RenderLayer::clearSnapshottedScrollOffsetForAnchorPositioning): Deleted.
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForAnchorPositioning const):
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::isTransformed const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldUseTransformFromContainer const):

Update terminology and tweak for new scroll adjustment system.

* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateAnchorPositionedAfterScroll):

Update anchor scroll adjustment call site.

* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderObjectInlines.h:
(WebCore::RenderObject::layoutContext const):

Allow non-const access to layoutContext.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setIsForceHidden):

Allow setting force-hidden to true or false.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::AnchorScrollSnapshot::AnchorScrollSnapshot):
(WebCore::AnchorScrollSnapshot::AnchorScrollSnapshot):
(WebCore::AnchorScrollSnapshot::adjustmentForCurrentScrollPosition):
(WebCore::AnchorScrollAdjuster::isEmpty const):
(WebCore::AnchorScrollAdjuster::AnchorScrollAdjuster):
(WebCore::AnchorScrollAdjuster::addSnapshot):
(WebCore::AnchorScrollAdjuster::addViewportSnapshot):
(WebCore::AnchorScrollAdjuster::adjustmentForViewport):
(WebCore::AnchorScrollAdjuster::accumulateAdjustments):
(WebCore::AnchorScrollAdjuster::setFallbackLimits):
(WebCore::AnchorScrollAdjuster::invalidateForScroller):

Implement new anchor scroll adjuster system.

(WebCore::Style::clearAnchorScrollSnapshots):
(WebCore::Style::AnchorPositionEvaluator::captureScrollSnapshots):
(WebCore::Style::AnchorPositionEvaluator::updateScrollAdjustments):

Rewrite anchor scrolling to use new anchor scroll adjuster system.

(WebCore::Style::offsetFromAncestorContainer):
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):
(WebCore::Style::computeInsetValue):

Incorporate scroll positions into anchor rect computation.

(WebCore::Style::AnchorPositionEvaluator::scrollOffsetFromAnchor): Deleted.
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets): Deleted.
(WebCore::Style::AnchorPositionEvaluator::updatePositionsAfterScroll): Deleted.

Remove old scroll adjustment system.

* Source/WebCore/style/AnchorPositionEvaluator.h:
(WebCore::AnchorScrollAdjuster::anchored const):
(WebCore::AnchorScrollAdjuster::mayNeedAdjustment const):
(WebCore::AnchorScrollAdjuster::isHidden const):
(WebCore::AnchorScrollAdjuster::setHidden):
(WebCore::AnchorScrollAdjuster::hasFallbackLimits const):
(WebCore::AnchorScrollAdjuster::exceedsFallbackLimits):

Implement new anchor scroll adjuster system.

* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Don&apos;t trigger invalidation due to scroll adjustments.

(WebCore::Style::Scope::updateAnchorPositioningStateAfterStyleResolution):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::updateForPositionVisibility):

Update anchor scroll adjustment call sites.

Canonical link: <a href="https://commits.webkit.org/298770@main">https://commits.webkit.org/298770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ff4ceac4ffced6ab427b4aeec3d8b0c7654db89

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116569 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36230 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122634 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67135 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d289a5f7-181f-41df-aa14-d42d62b51efd) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118458 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44818 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88527 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42962 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/83831827-85ce-4411-ad46-64aee2822b02) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119518 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29448 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104566 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68991 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28513 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22675 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66308 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98825 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22833 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125775 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32641 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97208 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43828 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100775 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97003 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24704 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42314 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20234 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39417 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43351 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48942 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42817 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46157 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44522 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -815,6 +815,44 @@ void LocalFrameViewLayoutContext::checkLayoutState()
+    });
+    if (WTF::notFound == index)
+        return { };
+    return &m_anchorScrollAdjusters[index];
```
Alternatively could use a `WeakHashMap<const RenderBox, AnchorScrollAdjuster>` for m_anchorScrollAdjusters

fantasai replied:
That was my first approach, actually, but it meant externalizing the storage for the anchor scroll adjusters, and I wanted to keep them inline. I think it's unlikely that a page will have many items in this list, so I'm not too worried about the performance of finding them.

anttijk replied:
Not sure what you mean by "externalizing". Vector uses non-inline heap storage by default, no different from HashMap. Also at this point it is nicer looking code (which an associative container allows) is much more important than a micro-optimization (if that was the goal) of dubious benefit.

fantasai replied:
Meaning, I'd have to allocate each adjuster individually on the heap and store pointers in hashmap, which creates a sort of memory management problem that the code then needs to deal with. From what I could tell, the getters in HashMap return by value, and I wanted access by reference...

fantasai replied:
Anyway, I suggest we land this as-is, and we can follow up and convert the storage type later if it seems appropriate. But after fussing a bit with the hash table approach, I think this is a better approach.

Packaging up the anchor together with its data into a single object is actually much nicer for the user; the Vector class's convenience methods are pretty easy to use for lookups; and I don't think we're losing anything on performance given the expected usage patterns.

anttijk commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -805,6 +805,9 @@ bool RenderBlock::simplifiedLayout()
     updateScrollInfoAfterLayout();
 
+    if (isOutOfFlowPositioned())
+        Style::AnchorPositionEvaluator::captureScrollSnapshots(*this);
```
Do we expect to need the snapshots for non-anchor positioned boxes out-of-flow boxes? I suppose position-visibility:no-overflow might need it.

fantasai replied:
Hm, no I think we would only need them for anchored boxes... regular positioned boxes don't run out of room if you scroll. Right?

(Fwiw, I put the out-of-flow conditional outside the function for two reasons: 1. avoids the extra function call for most boxes and 2. makes it clearer to readers of the caller that this is an out-of-flow thing they don't have to think about for other cases. I kept the more complicated "is it anchor positioned and do we need snapshots" checks inside the function to encapsulate those and keep it easy to call. We could alternatively make it part of an early return.)

anttijk replied:
I think you could just do Style::AnchorPositionEvaluator::isAnchorPositioned instead. It does the isOutOfFlowPositioned check first. That way it is clear that this is about anchor positioning only. The function itself can then ASSERT(isAnchorPositioned)

fantasai replied:
Done.

anttijk commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2734,6 +2735,8 @@ void RenderElement::layoutIfNeeded()
     }
     layout();
+    if (isOutOfFlowPositioned())
+        Style::AnchorPositionEvaluator::captureScrollSnapshots(downcast<RenderBox>(*this));
```
@alanbaradlay might have opinions about the best place in the code to do these calls.

alanbaradlay replied:
I do have opinions, but sadly that best place does not exist yet. this should do for now..

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.h:
```diff
@@ -42,16 +45,61 @@ namespace WebCore {
 enum CSSPropertyID : uint16_t;
 
+struct AnchorScrollSnapshot {
+    const RenderBox* m_scroller;
```
Please use `WeakPtr<const RenderBox>` or `CheckedPtr<const RenderBox>`. The latter will crash in case the RenderBox dies before the snapshot is deleted so requires correct cleanup code.

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.h:
```diff
@@ -42,16 +45,61 @@ namespace WebCore {
+private:
+    LayoutSize adjustmentForViewport(const RenderView&) const;
+
+    RenderBox* m_anchored;
```
Please use `WeakPtr<const RenderBox>` or `CheckedPtr<const RenderBox>`.

anttijk commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -173,6 +173,10 @@ void RenderBox::willBeDestroyed()
         if (!style().positionTryFallbacks().isEmpty())
             view().unregisterPositionTryBox(*this);
+        if (isOutOfFlowPositioned()) { // Keep this in sync with styleDidChange().
+            if (Style::AnchorPositionEvaluator::isLayoutTimeAnchorPositioned(style()) || style().usesAnchorFunctions())
```
These are equivalent to `if (Style::AnchorPositionEvaluator::isAnchorPositioned(style()))`

anttijk commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -408,7 +412,19 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+        && (Style::AnchorPositionEvaluator::isLayoutTimeAnchorPositioned(*oldStyle) || oldStyle->usesAnchorFunctions())
+        && !(Style::AnchorPositionEvaluator::isLayoutTimeAnchorPositioned(style()) || style().usesAnchorFunctions())) {
+        layoutContext().unregisterAnchorScrollAdjusterFor(*this);
+    }
```
These can use `Style::AnchorPositionEvaluator::isAnchorPositioned`

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1639,8 +1639,7 @@ void TreeResolver::updateForPositionVisibility(RenderStyle& style, const Styleab
     // FIXME: Implement via "visibility: force-hidden".
-    if (shouldHideAnchorPositioned())
-        style.setIsForceHidden();
+    style.setIsForceHidden(shouldHideAnchorPositioned());
```
Why this change? We should start the style building with the bit set to false. Which test fails without it?

fantasai replied:
Tests where the anchor box starts scrolled out of view, and is then scrolled into view.

anttijk replied:
Ok, I can check what is going on after landing.

anttijk reviewed (approved):
looks good

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -815,6 +815,44 @@ void LocalFrameViewLayoutContext::checkLayoutState()
+        return item.anchored() == &anchored;
+    });
+    if (WTF::notFound == index)
+        return { };
```
```
 if (index == WTF::notFound)
        return nullptr;
```

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -815,6 +815,44 @@ void LocalFrameViewLayoutContext::checkLayoutState()
+
+void LocalFrameViewLayoutContext::registerAnchorScrollAdjuster(AnchorScrollAdjuster&& scrollAdjuster)
+{
+
```
Blank line

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -815,6 +815,44 @@ void LocalFrameViewLayoutContext::checkLayoutState()
+    m_anchorScrollAdjusters.removeFirstMatching([&](auto& item) {
+        return item.anchored() == &anchored;
+    });
+    ASSERT(!m_anchorScrollAdjusters.containsIf([&](auto& item) {
```
You should use a data structure that doesn't allow duplicates (HashSet, HashMap) instead of asserting.

fantasai replied:
That was my first approach, but I wanted to keep these items inline in memory rather than externalizing them individually on the heap, and it doesn't seem those data types allow for reference access to inline data. (?)

fantasai replied:
(See follow-up discussion in https://github.com/WebKit/WebKit/pull/49318#discussion_r2272557988)

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -164,6 +165,12 @@ class LocalFrameViewLayoutContext final : public CanMakeCheckedPtr<LocalFrameVie
     void deleteDetachedRenderersNow() const { m_detachedRendererList.clear(); }
 
+    Vector<AnchorScrollAdjuster>& anchorScrollAdjusters() { return m_anchorScrollAdjusters; }
+    AnchorScrollAdjuster* anchorScrollAdjusterFor(const RenderBox& anchored);
```
Can this be a const function?

fantasai replied:
Hmm, good point. I'll give it a try. :)

fantasai replied:
It compiles. :) Yay.

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -173,6 +173,10 @@ void RenderBox::willBeDestroyed()
             view().unregisterAnchor(*this);
         if (!style().positionTryFallbacks().isEmpty())
             view().unregisterPositionTryBox(*this);
+        if (isOutOfFlowPositioned()) { // Keep this in sync with styleDidChange().
```
Instead of the comment, can we just share code?

fantasai replied:
Not really, the conditions are different: one is "are we this kind of thing" and the other is "did we change to or away from this kind of thing", so we can't just encapsulate it all into a single function call.

fantasai replied:
We can make the conditionals a bit more similar, but I think it's good to keep the comment so people don't add conditions that take it out of sync.

fantasai replied:
I did some clean-up to keep the constructions more parallel. But I kept the comment in place, because it still seems like something that could trip someone up.

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -408,7 +412,19 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+        layoutContext().unregisterAnchorScrollAdjusterFor(*this);
+        if (layer())
+            layer()->clearAnchorScrollAdjustment();
+    } else if (oldStyle && isOutOfFlowPositioned() // Keep this in sync with willBeDestroyed().
```
Same question about the comment.

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -52,7 +53,162 @@
+AnchorScrollAdjuster::AnchorScrollAdjuster(RenderBox& anchored, const RenderBoxModelObject& defaultAnchor)
+    : m_anchored(&anchored)
+{
+    const RenderStyle& style = anchored.style();
```
This can just be `auto& style` (the compiler will deduce constness).

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -267,6 +474,19 @@ LayoutRect AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock(C
+        // Handle fixed positioning x scrolling anchor.
+        bool isFixedAnchor = false;
+        for (const RenderElement* box = anchorBox.ptr(); box && box != &containingBlock; box = box->container()) {
+            if (box->isFixedPositioned()) {
```
Note that this only consults style, and not whether some container (e.g. a transform or filter) prevents fixedpos behavior). For that you'd consult `zlayer->behavesAsFixed()`.

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.h:
```diff
@@ -42,16 +45,61 @@ namespace WebCore {
+    inline void addSnapshot(const RenderBox& scroller);
+    inline void addViewportSnapshot(const RenderView&);
+
+    void setFallbackLimits(const RenderBox& anchored);
```
Weird to have a "set" that doesn't take a limits argument. Maybe `computeFallbackLimits()`

fantasai replied:
It also flips the fallback limits on, not just computing what they are. The focus of the function is on setting them, really, not so much on computing them. (They're not normally set.) It computes from the RenderBox as a convenience.

fantasai committed:
Make anchor positioning fallbacks respond to scrolling [css-anchor-position-1]
https://bugs.webkit.org/show_bug.cgi?id=295003
rdar://154349365

Reviewed by Antti Koivisto.

This patch rewrites are scroll handling for anchor positioning to incorporate
the current scroll position into layout so that anchored boxes can respond to
space available at the current scroll position.

In order to maintain responsiveness to scrolling, it caches scroll position
snapshots with each layout, which allows us to rapidly calculate adjustments
o the anchored box's scroll adjustment transform on changes to the scroll
position outside of layout as well as to trigger more invasive invalidation
when necessary in response to scroll position changes, e.g. for fallbacks.

* LayoutTests/TestExpectations:

Pass more tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-chained-002-expected.html:

Fix test reference.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-js-expose-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt:

Pass more tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-011-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-scrolling-002.tentative-expected.txt:

Rebaseline.

* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:

Mark WK1 tests as still failing, to debug later per Antti's instructions.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::updateAnchorPositionedAfterScroll):

Update anchor scroll adjustment call site.

* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::anchorScrollAdjusterFor):
(WebCore::LocalFrameViewLayoutContext::registerAnchorScrollAdjuster):
(WebCore::LocalFrameViewLayoutContext::unregisterAnchorScrollAdjusterFor):
(WebCore::LocalFrameViewLayoutContext::invalidateAnchorDependenciesForScroller):
(WebCore::LocalFrameViewLayoutContext::removeScrollerFromAnchorScrollAdjusters):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:

Set up storage for anchor scroll adjuster objects.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):

Update anchor scroll adjustment call site.

* Source/WebCore/platform/graphics/LayoutPoint.h:
(WebCore::LayoutPointLimits::clamp):
(WebCore::LayoutPointLimits::fits):
(WebCore::LayoutPointLimits::distance):
* Source/WebCore/platform/graphics/LayoutSize.h:
(WebCore::LayoutSizeLimits::clamp):
(WebCore::LayoutSizeLimits::fits):
(WebCore::LayoutSizeLimits::distance):

Introduce some convenience objects / methods.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::captureAnchorGeometry):

Adjust call parameters to computeAnchorRectRelativeToContainingBlock().

* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::originalContainingRange const):
(WebCore::PositionedLayoutConstraints::insetModifiedContainingRange const):

Export more info.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::simplifiedLayout):
(WebCore::RenderBlock::layoutOutOfFlowBox):

Call scroll snapshot capture method after layout.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::willBeDestroyed):
(WebCore::RenderBox::styleDidChange):

Manage anchor scroll adjuster memory allocation and invalidation.

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::layoutIfNeeded):

Call scroll snapshot capture method after layout.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransformFromStyle const):
(WebCore::RenderLayer::setAnchorScrollAdjustment):
(WebCore::RenderLayer::clearAnchorScrollAdjustment):
(WebCore::RenderLayer::setSnapshottedScrollOffsetForAnchorPositioning): Deleted.
(WebCore::RenderLayer::clearSnapshottedScrollOffsetForAnchorPositioning): Deleted.
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForAnchorPositioning const):
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::isTransformed const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldUseTransformFromContainer const):

Update terminology and tweak for new scroll adjustment system.

* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateAnchorPositionedAfterScroll):

Update anchor scroll adjustment call site.

* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderObjectInlines.h:
(WebCore::RenderObject::layoutContext const):

Allow non-const access to layoutContext.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setIsForceHidden):

Allow setting force-hidden to true or false.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::AnchorScrollSnapshot::AnchorScrollSnapshot):
(WebCore::AnchorScrollSnapshot::AnchorScrollSnapshot):
(WebCore::AnchorScrollSnapshot::adjustmentForCurrentScrollPosition):
(WebCore::AnchorScrollAdjuster::isEmpty const):
(WebCore::AnchorScrollAdjuster::AnchorScrollAdjuster):
(WebCore::AnchorScrollAdjuster::addSnapshot):
(WebCore::AnchorScrollAdjuster::addViewportSnapshot):
(WebCore::AnchorScrollAdjuster::adjustmentForViewport):
(WebCore::AnchorScrollAdjuster::accumulateAdjustments):
(WebCore::AnchorScrollAdjuster::setFallbackLimits):
(WebCore::AnchorScrollAdjuster::invalidateForScroller):

Implement new anchor scroll adjuster system.

(WebCore::Style::clearAnchorScrollSnapshots):
(WebCore::Style::AnchorPositionEvaluator::captureScrollSnapshots):
(WebCore::Style::AnchorPositionEvaluator::updateScrollAdjustments):

Rewrite anchor scrolling to use new anchor scroll adjuster system.

(WebCore::Style::offsetFromAncestorContainer):
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):
(WebCore::Style::computeInsetValue):

Incorporate scroll positions into anchor rect computation.

(WebCore::Style::AnchorPositionEvaluator::scrollOffsetFromAnchor): Deleted.
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets): Deleted.
(WebCore::Style::AnchorPositionEvaluator::updatePositionsAfterScroll): Deleted.

Remove old scroll adjustment system.

* Source/WebCore/style/AnchorPositionEvaluator.h:
(WebCore::AnchorScrollAdjuster::anchored const):
(WebCore::AnchorScrollAdjuster::mayNeedAdjustment const):
(WebCore::AnchorScrollAdjuster::isHidden const):
(WebCore::AnchorScrollAdjuster::setHidden):
(WebCore::AnchorScrollAdjuster::hasFallbackLimits const):
(WebCore::AnchorScrollAdjuster::exceedsFallbackLimits):

Implement new anchor scroll adjuster system.

* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Don't trigger invalidation due to scroll adjustments.

(WebCore::Style::Scope::updateAnchorPositioningStateAfterStyleResolution):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::updateForPositionVisibility):

Update anchor scroll adjustment call sites.

Canonical link: https://commits.webkit.org/298770@main
